⚡️ preload entries and bulk insert in JUMP_SMOOTHING

JGreenlee opened the pull request:
### ⚡️ preload entries and bulk insert in JUMP_SMOOTHING

In JUMP_SMOOTHING, we iterate over all sections and look at the locations in them to identify jumps. Instead of making separate 'find' queries for each section, we can query all locations upfront and filter by time range to look at specific sections (similar to what we have done for trip segmentation and section segmentation: https://github.com/e-mission/e-mission-server/pull/1036 https://github.com/e-mission/e-mission-server/pull/1014)

Additionally, we can reduce the number of 'insert' queries by keeping a list of the "entries to be created" and doing a bulk_insert once we are finished iterating

Overall this reduces the number of DB queries from O(n) where n is # of sections, to O(1)
For a moderately large day of travel this is a reduction of about 10x:
https://github.com/e-mission/e-mission-server/pull/1042#issuecomment-2776979556

<hr>

This currently causes tests to fail but I don't know why. will update later

JGreenlee commented:
This would appear to be a straightforward change, but I have not been able to get tests to pass even after several hours of debugging.

The distances of the cleaned trips are shorter than expected, even though the created `analysis/smoothing` entries appear identical and logs appear identical. I have also added additional log statements in several places, but I have not been able to isolate what is causing it to behave differently

JGreenlee commented:
Working backwards and digging through the CLEAN_RESAMPLING stage, I found that the points flagged in the `analysis/smoothing` entries are sometimes not filtered out during `remove_outliers`. This explains the differing distances.

https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/analysis/intake/cleaning/clean_and_resample.py#L771-L784

But this should not be the case, because as I noted:

> the created `analysis/smoothing` entries appear identical

They **appear** identical, the only difference being the _id values of the `deleted_points`.
I had made sure they were referring to the same point by looking at the ts and lat/lon, and I didn't think much of the differing _ids since those are random and often differ from one run to another.

It turns out that the differing _ids **is** a problem because those _ids come from `shankari_2016-06-20`, and the reason I'm seeing them mixed up is because there are duplicate entries.
There are 2 `background/filtered_location` entries with ts=`1466436690.661`. They have all the same readings but different `write_ts`.

https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/tests/data/real_examples/shankari_2016-06-20#L2169-L2182

https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/tests/data/real_examples/shankari_2016-06-20#L38184-L38197

Something about my changes to preload the locations in JUMP_SMOOTHING has made it so that the location dataframe has `576810cb4a9bba50e3fa1d74` instead of `57681ee64a9bba50e3fa1ffe` (even though they refer to the same point!).
`576810cb4a9bba50e3fa1d74` gets flagged and added to the `deleted_points`

In CLEAN_RESAMPLING `remove_outliers`, we read locations again and this time we get `57681ee64a9bba50e3fa1ffe`, not `576810cb4a9bba50e3fa1d74`.
We look for `576810cb4a9bba50e3fa1d74` to filter it out, but it's not there and nothing happens.

JGreenlee commented:
@shankari Are duplicate location entries something we expect to see? Have there been previous instances?

shankari commented:
Great job digging into this! I can assume that it was very painful!

> Are duplicate location entries something we expect to see? Have there been previous instances?

I think that this is essentially the same as the OOO entries; there are two entries with the same `data.ts` but different `metadata.write_ts` => we got an "older" point at the later `metadata.write_ts`, so when sorted by `metadata.write_ts`, the points will appear to be out of order.

But we should have marked those as invalid in the database in the trip segmentation step 
https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/analysis/intake/segmentation/trip_segmentation.py#L103

and so they should not have been queried in this step
https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/storage/timeseries/builtin_timeseries.py#L134

So my questions are:
- were these locations identified as OOO in the trip segmentation step?
- were they marked as invalid in the database?
- if so, why were they read in the jump_smoothing stage?

shankari commented:
@JGreenlee also, it looks like the tests are passing now; what did you change?

JGreenlee commented:
> @JGreenlee also, it looks like the tests are passing now; what did you change?

Unfortunately, I don't exactly know. I was surprised to see them pass.

They failed yesterday due to a typo, which is not the issue I have been painfully investigating.

That issue only happens locally under certain circumstances. I just discovered that it goes away when I purge the DB, and it comes back if I run intake and then try to run tests. I can reproduce it consistently, but it's still a little mysterious.
It seems to only occur when the DB is in a certain state – and even when the DB is in that state, it only occurs on this branch, not `master`

JGreenlee commented:
The effects of this change look promising, but I don't think it is safe to merge until I get to the bottom of the bug.

Although it's not happening in the CI, the fact that I can get it to happen locally on this branch but not on master does suggest it is a regression

<img width="816" alt="image" src="https://github.com/user-attachments/assets/d7a6bf74-23f6-49e5-9817-07bd797f4bf4" />

<img width="827" alt="image" src="https://github.com/user-attachments/assets/111f57c2-a0e6-431a-9ffb-a5fac9d8cf03" />
<img width="814" alt="image" src="https://github.com/user-attachments/assets/d8c1d84d-91e8-4850-98fe-14e1e9dff553" />

<hr>

shankari commented:
> But if step (3) is omitted or uses a different day, it doesn't happen:

This sort-of makes sense, since you are running the Jun20 test (`-k testJun20`) and so the data from `2016-07-25` won't affect it.

I also note that when we are loading the entries, we delete the id by default

```
    for i, entry in enumerate(entries):
        entry["user_id"] = override_uuid
        if not args.retain:
            del entry["_id"]
        if args.verbose is not None and i % args.verbose == 0:
            print("About to save %s" % entry)
        munged_entries.append(entry)
```

So when the test loads the same day's data, we get two copies.

So the questions are:
- how does this work on master?
- why are these not ignored?
   - either through the OOO check + invalid marking OR
   - through the dedup check
   https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/storage/timeseries/builtin_timeseries.py#L318

JGreenlee commented:
I added a log statement here to help me see what's going on:

```diff
diff --git a/emission/storage/timeseries/builtin_timeseries.py b/emission/storage/timeseries/builtin_timeseries.py
index f8fac871..cf8da391 100644
--- a/emission/storage/timeseries/builtin_timeseries.py
+++ b/emission/storage/timeseries/builtin_timeseries.py
@@ -315,6 +315,8 @@ class BuiltinTimeSeries(esta.TimeSeries):
             dedup_check_list = [item for item in ecwe.Entry.get_dedup_list(key)
                                 if item in df.columns] + ["metadata_write_ts"]
             numeric_check_list = [col for col in dedup_check_list if df[col].dtype != 'object']
+            if 'ts' in df.columns:
+                print(f"df before de-duping:\n{df[['_id', 'ts']]}")
             deduped_df = df.drop_duplicates(subset=numeric_check_list)
             logging.debug("After de-duping, converted %s points to %s " %
                           (len(df), len(deduped_df)))
```

On `optimize_db_jump_smoothing`, we query the entire range which gives:
<details>

```
df before de-duping:
                          _id            ts
0    576810cb4a9bba50e3fa1d56 1466436483.39
1    57681ee64a9bba50e3fa1fe0 1466436483.39
2    576810cb4a9bba50e3fa1d59 1466436506.53
3    57681ee64a9bba50e3fa1fe3 1466436506.53
4    576810cb4a9bba50e3fa1d5c 1466436515.23
5    57681ee64a9bba50e3fa1fe6 1466436515.23
6    576810cb4a9bba50e3fa1d5f 1466436546.03
7    57681ee64a9bba50e3fa1fe9 1466436546.03
8    576810cb4a9bba50e3fa1d62 1466436553.96
9    57681ee64a9bba50e3fa1fec 1466436553.96
10   576810cb4a9bba50e3fa1d65 1466436582.63
11   57681ee64a9bba50e3fa1fef 1466436582.63
12   576810cb4a9bba50e3fa1d68 1466436597.35
13   57681ee64a9bba50e3fa1ff2 1466436597.35
14   576810cb4a9bba50e3fa1d6b 1466436619.29
15   57681ee64a9bba50e3fa1ff5 1466436619.29
16   576810cb4a9bba50e3fa1d6d 1466436628.54
17   57681ee64a9bba50e3fa1ff7 1466436628.54
18   576810cb4a9bba50e3fa1d70 1466436654.78
19   57681ee64a9bba50e3fa1ffa 1466436654.78
20   576810cb4a9bba50e3fa1d72 1466436663.99
21   57681ee64a9bba50e3fa1ffc 1466436663.99
22   576810cb4a9bba50e3fa1d74 1466436690.66
23   57681ee64a9bba50e3fa1ffe 1466436690.66
24   576810cb4a9bba50e3fa1d78 1466436698.89
25   57681ee64a9bba50e3fa2002 1466436698.89
26   576810cb4a9bba50e3fa1d7a 1466436730.09
27   57681ee64a9bba50e3fa2004 1466436730.09
28   57681ee64a9bba50e3fa2008 1466436752.66
29   576810cb4a9bba50e3fa1d7e 1466436752.66
30   576810cb4a9bba50e3fa1d80 1466436784.46
31   57681ee64a9bba50e3fa200a 1466436784.46
32   576810cb4a9bba50e3fa1d86 1466436816.26
33   57681ee64a9bba50e3fa2010 1466436816.26
34   576810cb4a9bba50e3fa1d88 1466436848.08
35   57681ee64a9bba50e3fa2012 1466436848.08
36   576810cb4a9bba50e3fa1d8c 1466436863.97
37   57681ee64a9bba50e3fa2016 1466436863.97
38   576810cb4a9bba50e3fa1d8e 1466436899.03
39   57681ee64a9bba50e3fa2018 1466436899.03
40   576810cb4a9bba50e3fa1d92 1466436959.37
41   57681ee64a9bba50e3fa201c 1466436959.37
42   576810cb4a9bba50e3fa1d95 1466436986.27
43   57681ee64a9bba50e3fa201f 1466436986.27
44   57681ee64a9bba50e3fa2028 1466437082.80
45   576810cb4a9bba50e3fa1d9e 1466437082.80
46   57681ee64a9bba50e3fa202a 1466437098.37
47   576810cb4a9bba50e3fa1da0 1466437098.37
48   57681ee64a9bba50e3fa202d 1466437142.92
49   576810cb4a9bba50e3fa1da3 1466437142.92
50   57681ee64a9bba50e3fa2030 1466437170.29
51   576810cb4a9bba50e3fa1da6 1466437170.29
52   57681ee64a9bba50e3fa2035 1466437227.57
53   576810cb4a9bba50e3fa1dab 1466437227.57
54   57681ee64a9bba50e3fa2039 1466437275.86
55   576810cb4a9bba50e3fa1daf 1466437275.86
56   576810cb4a9bba50e3fa1db2 1466437303.76
57   57681ee64a9bba50e3fa203c 1466437303.76
58   576810cb4a9bba50e3fa1db5 1466437326.66
59   57681ee64a9bba50e3fa203f 1466437326.66
60   576810cc4a9bba50e3fa1db8 1466437335.54
61   57681ee64a9bba50e3fa2042 1466437335.54
62   576810cc4a9bba50e3fa1dbd 1466437402.36
63   57681ee64a9bba50e3fa2047 1466437402.36
64   576810cc4a9bba50e3fa1dc2 1466437462.42
65   57681ee74a9bba50e3fa204c 1466437462.42
66   57681ee74a9bba50e3fa204e 1466437492.49
67   576810cc4a9bba50e3fa1dc4 1466437492.49
68   57681ee74a9bba50e3fa2057 1466437531.59
69   57681ee74a9bba50e3fa205d 1466437585.14
70   57681ee74a9bba50e3fa205f 1466437615.13
71   57681ee74a9bba50e3fa2062 1466437625.05
72   57681ee74a9bba50e3fa2064 1466437645.21
73   57681ee74a9bba50e3fa2067 1466437657.63
74   57681ee74a9bba50e3fa206c 1466437693.60
75   57681ee74a9bba50e3fa206e 1466437706.28
76   57681ee74a9bba50e3fa2071 1466437729.21
77   57681ee74a9bba50e3fa2074 1466437769.01
78   57681ee74a9bba50e3fa2077 1466437804.85
79   57681ee74a9bba50e3fa207a 1466437817.54
80   57681ee74a9bba50e3fa207d 1466437860.75
81   57681ee74a9bba50e3fa2080 1466437896.47
82   57681ee74a9bba50e3fa2083 1466437928.91
83   57681ee74a9bba50e3fa2086 1466437944.19
84   57681ee74a9bba50e3fa2089 1466437980.02
85   57681ee74a9bba50e3fa208b 1466437986.31
86   57681ee74a9bba50e3fa208f 1466438011.58
87   57681ee74a9bba50e3fa2091 1466438022.96
88   57681ee74a9bba50e3fa2093 1466438041.59
89   57681ee74a9bba50e3fa2096 1466438063.27
90   57681ee74a9bba50e3fa2098 1466438073.17
91   57681ee74a9bba50e3fa209b 1466438099.14
92   57681ee74a9bba50e3fa209d 1466438104.70
93   57681ee74a9bba50e3fa20a1 1466438136.00
94   57681ee74a9bba50e3fa20a3 1466438163.12
95   57681ee74a9bba50e3fa20a6 1466438172.08
96   576883c488f6630bbd0faa06 1466461675.31
97   576876b54a9bba50e3faac36 1466461683.13
98   576876b54a9bba50e3faac3a 1466461699.75
99   576876b54a9bba50e3faac3c 1466461714.95
100  576876b54a9bba50e3faac3f 1466461735.74
101  576876b54a9bba50e3faac41 1466461746.79
102  576876b54a9bba50e3faac44 1466461771.70
103  576876b54a9bba50e3faac46 1466461778.63
104  576876b54a9bba50e3faac49 1466461806.97
105  576876b54a9bba50e3faac4d 1466461848.02
106  576876b54a9bba50e3faac50 1466461889.89
107  576876b54a9bba50e3faac52 1466461917.08
108  576876b54a9bba50e3faac54 1466461945.33
109  576876b54a9bba50e3faac57 1466461966.38
110  576876b54a9bba50e3faac59 1466461973.21
111  576876b54a9bba50e3faac5c 1466462030.32
112  576876b54a9bba50e3faac5e 1466462037.00
113  576876b54a9bba50e3faac62 1466462074.70
114  576876b54a9bba50e3faac65 1466462087.15
115  576876b54a9bba50e3faac67 1466462118.58
116  576876b54a9bba50e3faac6f 1466462234.85
117  576876b54a9bba50e3faac72 1466462253.18
118  576876b54a9bba50e3faac74 1466462266.68
119  576876b54a9bba50e3faac77 1466462293.08
120  576876b54a9bba50e3faac7a 1466462310.14
121  576876b54a9bba50e3faac7c 1466462331.56
122  576876b54a9bba50e3faac7f 1466462342.26
123  576876b54a9bba50e3faac81 1466462367.20
124  576876b54a9bba50e3faac87 1466462403.65
125  576876b54a9bba50e3faac8b 1466462410.41
126  576876b54a9bba50e3faac8e 1466462452.71
127  576876b54a9bba50e3faac91 1466462459.60
128  576876b54a9bba50e3faac94 1466462493.14
129  576876b54a9bba50e3faac98 1466462529.69
130  576876b54a9bba50e3faac9a 1466462545.04
131  576876b54a9bba50e3faac9e 1466462565.67
132  576876b54a9bba50e3faaca0 1466462576.57
133  576876b54a9bba50e3faaca3 1466462600.73
134  576876b54a9bba50e3faaca6 1466462608.23
135  576885474a9bba50e3faaf5a 1466463008.66
136  576885474a9bba50e3faaf5d 1466463028.22
137  576885474a9bba50e3faaf5f 1466463040.47
138  576885474a9bba50e3faaf61 1466463063.16
139  576885474a9bba50e3faaf64 1466463072.28
140  576885474a9bba50e3faaf67 1466463099.83
141  576885474a9bba50e3faaf6b 1466463147.68
142  576885474a9bba50e3faaf6f 1466463182.14
143  576885474a9bba50e3faaf72 1466463199.52
144  576885474a9bba50e3faaf74 1466463220.25
145  576885474a9bba50e3faaf77 1466463231.32
146  576885474a9bba50e3faaf79 1466463261.24
147  576885474a9bba50e3faaf7d 1466463303.18
148  576885474a9bba50e3faaf81 1466463344.34
149  576885474a9bba50e3faaf84 1466463358.56
150  576885474a9bba50e3faaf86 1466463384.33
151  576885474a9bba50e3faaf8a 1466463425.23
152  576885474a9bba50e3faaf8e 1466463468.55
153  576885474a9bba50e3faaf93 1466463508.51
154  576885474a9bba50e3faaf95 1466463517.64
155  576885474a9bba50e3faaf9a 1466463550.27
156  576885474a9bba50e3faaf9d 1466463586.83
157  576885474a9bba50e3faafa2 1466463643.80
158  576885474a9bba50e3faafa6 1466463680.19
159  576885474a9bba50e3faafa9 1466463685.96
160  576885474a9bba50e3faafad 1466463736.52
161  576885474a9bba50e3faafb0 1466463772.10
162  576885474a9bba50e3faafb2 1466463783.09
163  576885484a9bba50e3faafb7 1466463803.90
164  576885484a9bba50e3faafb9 1466463821.13
165  576885484a9bba50e3faafbc 1466463835.71
166  576885484a9bba50e3faafbe 1466463864.24
167  576885484a9bba50e3faafc2 1466463901.10
168  576885484a9bba50e3faafc5 1466463921.62
169  576885484a9bba50e3faafc8 1466463953.40
170  576885484a9bba50e3faafca 1466463962.15
171  576885484a9bba50e3faafcd 1466463985.21
172  576885484a9bba50e3faafcf 1466464013.98
173  576885484a9bba50e3faafd7 1466464321.61
174  576885484a9bba50e3faafda 1466464338.56
175  576885484a9bba50e3faafdc 1466464369.66
176  576885484a9bba50e3faafde 1466464376.89
177  576885484a9bba50e3faafe1 1466464404.99
178  576885484a9bba50e3faafe4 1466464423.18
179  576885484a9bba50e3faafe6 1466464441.56
180  576885484a9bba50e3faafe9 1466464453.21
181  576885484a9bba50e3faafeb 1466464477.22
182  576885484a9bba50e3faafee 1466464483.23
183  576885484a9bba50e3faaff1 1466464516.37
184  576894bf4a9bba50e3fab302 1466466669.68
185  576894bf4a9bba50e3fab305 1466466680.96
186  576894bf4a9bba50e3fab307 1466466705.55
187  576894bf4a9bba50e3fab309 1466466715.15
188  576894bf4a9bba50e3fab30d 1466466740.42
189  576894bf4a9bba50e3fab30f 1466466750.03
190  576894bf4a9bba50e3fab311 1466466770.36
191  576894bf4a9bba50e3fab314 1466466784.88
192  576894bf4a9bba50e3fab316 1466466801.51
193  576894bf4a9bba50e3fab319 1466466820.58
194  576894bf4a9bba50e3fab31b 1466466832.67
195  576894bf4a9bba50e3fab31e 1466466855.24
196  576894bf4a9bba50e3fab321 1466466889.86
197  576894bf4a9bba50e3fab325 1466466900.36
198  576894bf4a9bba50e3fab327 1466466922.42
199  576894bf4a9bba50e3fab32b 1466466948.81
200  576894bf4a9bba50e3fab32d 1466466959.53
201  576894bf4a9bba50e3fab330 1466466979.33
202  576894bf4a9bba50e3fab332 1466466997.41
203  576894bf4a9bba50e3fab335 1466467009.31
204  576894bf4a9bba50e3fab337 1466467033.26
205  576894bf4a9bba50e3fab33a 1466467039.50
206  576894bf4a9bba50e3fab33d 1466467065.71
207  576894bf4a9bba50e3fab340 1466467081.43
208  576894bf4a9bba50e3fab342 1466467101.62
209  576894bf4a9bba50e3fab344 1466467111.42
210  576894bf4a9bba50e3fab347 1466467134.18
211  576894bf4a9bba50e3fab349 1466467142.48
212  576894bf4a9bba50e3fab34c 1466467166.68
213  576894bf4a9bba50e3fab34e 1466467179.92
214  576894bf4a9bba50e3fab351 1466467203.57
215  576894bf4a9bba50e3fab354 1466467214.90
216  576894bf4a9bba50e3fab356 1466467237.14
217  576894bf4a9bba50e3fab35a 1466467273.14
218  576894bf4a9bba50e3fab35d 1466467278.33
219  576894bf4a9bba50e3fab35f 1466467309.20
220  576894bf4a9bba50e3fab362 1466467336.70
221  576894bf4a9bba50e3fab365 1466467341.75
222  576894bf4a9bba50e3fab367 1466467366.72
223  576894bf4a9bba50e3fab369 1466467376.64
224  576894bf4a9bba50e3fab36c 1466467403.94
225  576894bf4a9bba50e3fab36f 1466467415.72
226  576894bf4a9bba50e3fab371 1466467439.98
227  576894bf4a9bba50e3fab373 1466467452.31
228  576894bf4a9bba50e3fab376 1466467475.53
229  576894bf4a9bba50e3fab378 1466467484.76
230  576894bf4a9bba50e3fab37b 1466467510.75
231  576894bf4a9bba50e3fab37d 1466467519.65
232  576894bf4a9bba50e3fab380 1466467540.77
233  576894bf4a9bba50e3fab382 1466467555.33
234  576894bf4a9bba50e3fab386 1466467590.22
235  576894bf4a9bba50e3fab389 1466467610.56
236  576894bf4a9bba50e3fab38c 1466467631.75
237  576894bf4a9bba50e3fab38e 1466467639.85
238  576894bf4a9bba50e3fab390 1466467672.37
239  576894bf4a9bba50e3fab392 1466467677.96
240  576894bf4a9bba50e3fab396 1466467720.81
241  576894bf4a9bba50e3fab399 1466467732.55
242  576894bf4a9bba50e3fab39b 1466467756.07
243  576894bf4a9bba50e3fab39e 1466467762.51
244  576894bf4a9bba50e3fab3a1 1466467793.71
245  576894bf4a9bba50e3fab3a6 1466467865.67
246  576894bf4a9bba50e3fab3a8 1466467900.21
247  576894bf4a9bba50e3fab3ab 1466467929.74
248  576894bf4a9bba50e3fab3ad 1466467939.10
249  576894bf4a9bba50e3fab3af 1466467959.77
250  576894bf4a9bba50e3fab3b2 1466467974.19
251  576894bf4a9bba50e3fab3b4 1466467991.28
252  576894bf4a9bba50e3fab3b7 1466468012.44
253  576894bf4a9bba50e3fab3b9 1466468022.98
254  576894bf4a9bba50e3fab3bc 1466468051.96
255  576894bf4a9bba50e3fab3bf 1466468087.63
```
</details>

The relevant part is 
```
22   576810cb4a9bba50e3fa1d74 1466436690.66
23   57681ee64a9bba50e3fa1ffe 1466436690.66
```

On `master`, we only query over that section, which is `1466436483.395 -> 1466437275.856` and that gives:

<details>
```
df before de-duping:
                         _id            ts
0   57681ee64a9bba50e3fa1fe0 1466436483.39
1   576810cb4a9bba50e3fa1d56 1466436483.39
2   57681ee64a9bba50e3fa1fe3 1466436506.53
3   576810cb4a9bba50e3fa1d59 1466436506.53
4   57681ee64a9bba50e3fa1fe6 1466436515.23
5   576810cb4a9bba50e3fa1d5c 1466436515.23
6   57681ee64a9bba50e3fa1fe9 1466436546.03
7   576810cb4a9bba50e3fa1d5f 1466436546.03
8   57681ee64a9bba50e3fa1fec 1466436553.96
9   576810cb4a9bba50e3fa1d62 1466436553.96
10  57681ee64a9bba50e3fa1fef 1466436582.63
11  576810cb4a9bba50e3fa1d65 1466436582.63
12  57681ee64a9bba50e3fa1ff2 1466436597.35
13  576810cb4a9bba50e3fa1d68 1466436597.35
14  57681ee64a9bba50e3fa1ff5 1466436619.29
15  576810cb4a9bba50e3fa1d6b 1466436619.29
16  57681ee64a9bba50e3fa1ff7 1466436628.54
17  576810cb4a9bba50e3fa1d6d 1466436628.54
18  57681ee64a9bba50e3fa1ffa 1466436654.78
19  576810cb4a9bba50e3fa1d70 1466436654.78
20  57681ee64a9bba50e3fa1ffc 1466436663.99
21  576810cb4a9bba50e3fa1d72 1466436663.99
22  57681ee64a9bba50e3fa1ffe 1466436690.66
23  576810cb4a9bba50e3fa1d74 1466436690.66
24  57681ee64a9bba50e3fa2002 1466436698.89
25  576810cb4a9bba50e3fa1d78 1466436698.89
26  57681ee64a9bba50e3fa2004 1466436730.09
27  576810cb4a9bba50e3fa1d7a 1466436730.09
28  57681ee64a9bba50e3fa2008 1466436752.66
29  576810cb4a9bba50e3fa1d7e 1466436752.66
30  57681ee64a9bba50e3fa200a 1466436784.46
31  576810cb4a9bba50e3fa1d80 1466436784.46
32  57681ee64a9bba50e3fa2010 1466436816.26
33  576810cb4a9bba50e3fa1d86 1466436816.26
34  57681ee64a9bba50e3fa2012 1466436848.08
35  576810cb4a9bba50e3fa1d88 1466436848.08
36  57681ee64a9bba50e3fa2016 1466436863.97
37  576810cb4a9bba50e3fa1d8c 1466436863.97
38  57681ee64a9bba50e3fa2018 1466436899.03
39  576810cb4a9bba50e3fa1d8e 1466436899.03
40  57681ee64a9bba50e3fa201c 1466436959.37
41  576810cb4a9bba50e3fa1d92 1466436959.37
42  57681ee64a9bba50e3fa201f 1466436986.27
43  576810cb4a9bba50e3fa1d95 1466436986.27
44  57681ee64a9bba50e3fa2028 1466437082.80
45  576810cb4a9bba50e3fa1d9e 1466437082.80
46  57681ee64a9bba50e3fa202a 1466437098.37
47  576810cb4a9bba50e3fa1da0 1466437098.37
48  57681ee64a9bba50e3fa202d 1466437142.92
49  576810cb4a9bba50e3fa1da3 1466437142.92
50  57681ee64a9bba50e3fa2030 1466437170.29
51  576810cb4a9bba50e3fa1da6 1466437170.29
52  57681ee64a9bba50e3fa2035 1466437227.57
53  576810cb4a9bba50e3fa1dab 1466437227.57
54  57681ee64a9bba50e3fa2039 1466437275.86
55  576810cb4a9bba50e3fa1daf 1466437275.86
```
</details>

This time, entry `1ffe` comes before entry `1d74`:
```
22  57681ee64a9bba50e3fa1ffe 1466436690.66
23  576810cb4a9bba50e3fa1d74 1466436690.66
```

<hr>

Why?

pandas `drop_duplicates` will always keep the first one since we did not set `keep`
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html

However, the order that we get the entries from mongodb is **not** consistent:
https://www.mongodb.com/docs/manual/reference/method/cursor.sort/#sort-consistency
> Sort Consistency
> MongoDB does not store documents in a collection in a particular order. When sorting on a field which contains duplicate values, documents containing those values may be returned in any order.
> 
> If consistent sort order is desired, include at least one field in your sort that contains unique values. The easiest way to guarantee this is to include the _id field in your sort query.

JGreenlee commented:
So the answer to

> * how does this work on master?

is "luck"

> why are these not ignored? ... through the OOO check + invalid marking

Looking into this next

JGreenlee commented:
> why are these not ignored? ... through the OOO check + invalid marking

When I said the duplicate points had different `write_ts`, I must have been mistaken. I may have been looking at the write_ts of the `background/location` entries (which there are also 2 of)

We only consider points as OOO if the ts decreases. If it stays the same, the diff is 0

I thought it might have something to do with my recent change 486a5440ddf15bf0760bb6c04b1199dd24eaea9b, but it looks like even before that we only checked for `ts.diff() < 0`

JGreenlee commented:
Since write_ts is the same, I'm questioning whether the duplicate entries actually came from the phone. Maybe they result from the way the timeline was exported for that day.
In fact, it looks like seems that the first 34 points of that day are duplicated, but everything from idx 68 onward is not duplicated
```
60   576810cc4a9bba50e3fa1db8 1466437335.54      1466437342.52
61   57681ee64a9bba50e3fa2042 1466437335.54      1466437342.52
62   576810cc4a9bba50e3fa1dbd 1466437402.36      1466437403.86
63   57681ee64a9bba50e3fa2047 1466437402.36      1466437403.86
64   576810cc4a9bba50e3fa1dc2 1466437462.42      1466437480.65
65   57681ee74a9bba50e3fa204c 1466437462.42      1466437480.65
66   57681ee74a9bba50e3fa204e 1466437492.49      1466437493.35
67   576810cc4a9bba50e3fa1dc4 1466437492.49      1466437493.35
68   57681ee74a9bba50e3fa2057 1466437531.59      1466437532.72
69   57681ee74a9bba50e3fa205d 1466437585.14      1466437586.05
70   57681ee74a9bba50e3fa205f 1466437615.13      1466437615.96
71   57681ee74a9bba50e3fa2062 1466437625.05      1466437625.84
72   57681ee74a9bba50e3fa2064 1466437645.21      1466437646.17
73   57681ee74a9bba50e3fa2067 1466437657.63      1466437658.44
```

Nonetheless, I could probably change the OOO detection to handle duplicates if we think that is necessary

We could also tack `_id` onto the end of our sort queries:

> If consistent sort order is desired, include at least one field in your sort that contains unique values. The easiest way to guarantee this is to include the _id field in your sort query.

JGreenlee commented:
#### 2cf5e6210ebea9be8a16338cd22b1b1b4074bd80 add '_id' as a secondary sort key for timeseries queries

> MongoDB sort queries produce inconsistent results when sorting on a field that has duplicate values. According to mongoDB docs, an easy way to prevent this is adding '_id' in the sort query, as '_id' is guaranteed to be unique
> 
> This resolves the issue being discussed in https://github.com/e-mission/e-mission-server/pull/1042#issuecomment-2779249958
> Tests pass when following those steps.

JGreenlee commented:
This time, tests failed in the workflows while passing locally

<hr>

It's because the workflow is using `pymongo` 4.3.3 while my local `emission` environment is on `4.4.1` and the allowed arguments to `sort` changed
https://pymongo.readthedocs.io/en/4.3.3/api/pymongo/cursor.html#pymongo.cursor.Cursor.sort
https://pymongo.readthedocs.io/en/4.4.1/api/pymongo/cursor.html#pymongo.cursor.Cursor.sort

Easy fix, but it does raise the question of why my local `emission` environment is using pymongo `4.4.1`

shankari commented:
wrt https://github.com/e-mission/e-mission-server/pull/1042#issuecomment-2779485835

I am fine with these changes; adding a secondary sort is not incorrect. However, it could affect the performance, particularly since we have added it for every query.

So I would still like to understand why this is an issue.
- If the points are identical (which they should be for the sort order to matter), why does it matter which order they are retrieved in?

Concretely, let us assume that we have two copies each of points p1 and p2 with times t1 and t2.
Let us assume further that their ids are i11, i12, i21 and i22

When we query MongoDB, we can get four combinations - (i11, i21), (i12, i21), (i11, i22), (i12, i22).
But given that i11 and i12 are identical except for the id, this should not matter; the time delta or any other computations should be identical.

Why does the order for completely identical entries matter?

JGreenlee commented:
> wrt https://github.com/e-mission/e-mission-server/pull/1042#issuecomment-2779485835
> 
> I am fine with these changes; adding a secondary sort is not incorrect. However, it could affect the performance, particularly since we have added it for every query.
> 
> So I would still like to understand why this is an issue.
> - If the points are identical (which they should be for the sort order to matter), why does it matter which order they are retrieved in?
> 
> Concretely, let us assume that we have two copies each of points p1 and p2 with times t1 and t2.
> Let us assume further that their ids are i11, i12, i21 and i22
> 
> When we query MongoDB, we can get four combinations - (i11, i21), (i12, i21), (i11, i22), (i12, i22).
> But given that i11 and i12 are identical except for the id, this should not matter; the time delta or any other computations should be identical.
> 
> Why does the order for completely identical entries matter?
> 

Because JUMP_SMOOTHING marks points for deletion by their ID 

In `JUMP_SMOOTHING`, we query locations and get (i11, i21). There's a jump and p2 is flagged as needing to be dropped. Its ID (i21) is stored in the `analysis/smoothing` entry for that section.

Later in `CLEANING_RESAMPLING`, we query locations and this time we get (i12, i22). We look at the `analysis/smoothing` entry for points to delete. i21 isn't there this time, so p2 never gets deleted.

shankari commented:
Ah great! Thank you for the clarification!

shankari commented on emission/analysis/intake/cleaning/location_smoothing.py:
```diff
@@ -109,11 +109,19 @@ def add_heading_change(points_df):
+        ts = esta.TimeSeries.get_time_series(user_id)
         sections_to_process = esda.get_entries(esda.RAW_SECTION_KEY, user_id,
                                                time_query)
+        time_query.timeType = "data.ts"
```
note to self: this is because before this change, we used to read the locations per section using `data.ts`

https://github.com/e-mission/e-mission-server/blob/823d47f062125a82e48276fcd542095bbb773d61/emission/storage/decorations/analysis_timeseries_queries.py#L103
